wfe/wfe.go

```diff
@@ -1627,6 +1639,61 @@ func (wfe *WebFrontEndImpl) makeChallenges(authz *core.Authorization, request *h
 	return nil
 }
 
+// validateReplacementOrder performs several sanity checks on the order to
+// determine if the order is a replacement of an existing order. If the order is
+// not a replacement or is a valid replacement, no problem will be returned.
+// Otherwise the caller will receive a problem.
+func (wfe *WebFrontEndImpl) validateReplacementOrder(newOrder *core.Order) *acme.ProblemDetails {
+	if newOrder == nil {
+		return acme.InternalErrorProblem("Order is nil")
+	}
+
+	if newOrder.Replaces == "" {
+		wfe.log.Printf("ARI: order %q is not a replacement\n", newOrder.ID)
+		return nil
+	}
+
+	certID, err := wfe.parseCertID(newOrder.Replaces)
+	if err != nil {
+		return acme.MalformedProblem(fmt.Sprintf("parsing ARI CertID failed: %s", err))
+	}
+
+	originalOrder, err := wfe.db.GetOrderByIssuedSerial(certID.ID)
+	if err != nil {
+		return acme.InternalErrorProblem(fmt.Sprintf("could not find an order for the given certificate: %s", err))
+	}
+
+	if originalOrder.IsReplaced {
+		return acme.Conflict(fmt.Sprintf("cannot indicate an order replaces certificate with serial %s, which already has a replacement order", certID.ID))
+	}
+
+	if originalOrder.AccountID != newOrder.AccountID {
+		return acme.UnauthorizedProblem("requester account did not request the certificate being replaced by this order")
+	}
+
+	// Servers SHOULD check that the identified certificate and the current New
+	// Order request correspond to the same ACME Account and share a
+	// preponderance of identifiers, and that the identified certificate has not
+	// already been marked as replaced by a different finalized Order. Servers
```

Comment:
This text [has been updated](https://github.com/aarongable/draft-acme-ari/pull/68) to 
> marked as replaced by a differed Order that is not "invalid"

So I think that [va.setOrderError](https://github.com/letsencrypt/pebble/blob/main/va/va.go#L212) should also call UpdateReplacedOrder to set the preceding order's IsReplaced field to _false_ when validation fails and this order becomes invalid.

Reply:
The VA is now a DB client to do accomplish this.